doc/contributing/maintaining-types-for-nodejs.md

```diff
@@ -28,6 +28,56 @@ The agreement was that the ideal flow would be as follows:
 * Automation within external type projects consumes the JSON and automatically
   generates a PR to add the API.
 
+## High level approach - development workflow
+
+The number of people using TypeScript with Node.js is significant enough
+that providing a good developer experience is important. While TypeScript
+is identified specifically, a secondary goal is that what we provide to improve
+development experience with TypeScript would apply to other type
+systems and transpiled languages as well.
+
+We have agreed that the approach will **NOT** include bundling TypeScript
+tools with Node.js but instead improve the developer experience for how
+those tools are installed/configured to work with Node.js.
+
+The high level developer experience we are working towards was captured in the
+[next-10 TypeScript mini-summit](https://github.com/nodejs/next-10/pull/150)
+and is as follows:
+
+1. When Node.js is started with an entry point that is not a file type that
+   Node.js recognizes, for example `node script.ts`, an informative error
+   message is printed that directs users to a webpage where they can
+   learn how to configure Node.js to support that file type.
+   * If the file was a TypeScript file, a TypeScript specific message with a
+     reference to a link on Nodejs.org specific on learning how to 
+     configure TypeScript will be provided.
+   * For other file types a generic message and shared webpage will be
+     used.
+2. Node.js gains support for loading configuration from a file. Most, if not
+   all, of the configuration supported by `NODE_OPTIONS` would be
+   supported in this file (which might be the `package.json` that lives
+   near the entry point file). The webpage with instructions would tell
+   users what configuration to put in this file to get Node.js to support
+   their file type.
+3. When Node.js is run with the correct configuration, either in a file or
+   `NODE_OPTIONS` or flags, the unknown file type is executed as expected. 
+
+Some additional specifics around the current approach include:
+
+* Loaders already provide a number of the components needed to
+  satisfy the requirements above. They already provide the Node.js
+  options that are needed to achieve many of the requirements above.
+* `package.json` as the location for the config is potentially a good
+  choice as Node.js already looks for it as part of startup.
+* The implementation chosen should allow for different configuration
+  in/for different environments/conditions like `prod`, `dev`, etc.
```

Comment:
I'm strongly -1 on adding the concept of NODE_ENV or analog to Node.js. However, the multiple environment is an issue.

This is often solved with supporting a "notable place config file" (possibly package.json) as well as a flag to specify said config file. Folks can support development with the default setting and override in prod, or viceversa.

Reply:
I think it’s best for this document to leave out implementation details of how this is achieved. I think there’s consensus that we wouldn’t use `NODE_ENV`, especially because it doesn’t make much sense to mix environment variables and files to define configuration. I don’t think it’s really necessary to specifically call out that we won’t be using `NODE_ENV`.

There’s already precedent for current Node supporting prod and dev (and other) configurations via the `--conditions` flag (https://nodejs.org/api/cli.html#-ccondition—conditionscondition) with those conditions appearing in `package.json` `"exports"`. I would assume that whatever we come with either uses that or is similar to it. But I don’t think we need to get into this level of detail here. This is just defining the goal.

Reply:
None of `--conditions` mentions staging, production or development.

How about:

> The implementation chosen should allow for being overridden via a different mechanism 
  for production-like environments.

Reply:
> None of --conditions mentions staging, production or development.

https://nodejs.org/api/cli.html#-ccondition%E2%80%94conditionscondition mentions development. The feature was created for this use case.

"Overridden via a different mechanism" is an implementation detail. The original text is fine. We need to support different configurations for different environments/conditions. That's enough.

Reply:
The original text mention specific environment names. As said, this is a deal breaker. NODE_ENV is one of the biggest mistakes we did as a community.

Reply:
```suggestion
  in/for different environments/conditions such as production
  versus development, or different types of hosted environments
  such as serverless vs traditional, etc.; Node.js would not make
  any recommendations or have any expectations as to what the
  separate configuration blocks should be named or what their
  purposes should be, just that a configuration file should have
  the ability to provide different configurations for user-defined
  conditions.
```

Reply:
That's ok, thx.